### PR TITLE
Jiayu --> Demi_create_no_records_state

### PR DESCRIPTION
**Pull request by this-journey**
Create a no records state.

# Description
Implements Phase 2 WBS 3.3.9: Adds no records state to records modal.

## Related PRS (if any):
none

## Main changes explained:
- Update records modal so that Update records and Purchase records both contain functionality for if no records are present
…

## How to test:
1. git pull
2. check into current branch
3. do npm install and npm run start:local to run this PR locally
4. Clear site data/cache in local storage
5. log in as any user
6. go to http://localhost:3000/bmdashboard/login to confirm your login credentials
7. go to http://localhost:3000/bmdashboard/materials
8. select "View" under "Purchases"
9. Verify that if no records are present modal reads "There are no purchase records"
10. select "View" under "Updates"
11. Verify that if no records are present modal reads "There are no updates to display"

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/93569791/fee3d265-9535-43bf-a575-60fc5dd54db5

## Note:
Use development branch for back end.

### CODE REVIEW

**Review by keyunhuangg — APPROVED** [review]
Check all 3 rows shown under updates and purchases that if no item is present then 'There are no purchase records.' under purchase record and 'There are no updates for this item.' under update records.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/124097392/cd7501cc-593f-4588-acb7-9070d871c2b9

**Comment by swathyjayaseelan** [comment]
Works as expected.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/13806470/d9c975de-2568-4202-b56c-deb0bbe36c46

**Review by 666saofeng — APPROVED** [review]
It worked as intended.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/62363348/1c96488e-b5a0-41a8-9c7d-21e16ca0c5d4

**Review by X1aoZhang99 — DISMISSED** [review]
Hi Demi,

I tested the PR with the admin role, I can see both "There are no purchase records" and "There are no updates to display." Nice work! 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113070229/d1edffb1-8d69-48ab-8035-62eb6755a7e4

**Review by peterson337 — DISMISSED** [review]
I tested your application, and everything worked as expected. Good job!

one comunnity

**Review by ClemarNunes — DISMISSED** [review]
I tested your PR and worked as expected.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/134466562/9019a432-67c3-446b-bbba-dd3c85e46867

**Review by mohabbasd — DISMISSED** [review]
Hello, I tested you PR and it works as intended, although I could see some updates done under someone else's name but purchase was empty. 
Screenshot 2024-02-17 at 11 13 27 PM
Screenshot 2024-02-17 at 11 13 19 PM

**Review by skaush21 — COMMENTED** [review]
* logged in as admin user, went to bmdashboard
* logged in again, went to bmdashboard/materials
* clicked on each of the views for update and purchase
* all button display correct output.
* for updates, when clicking on the 'view' for the ones that have items, a warning appears in console. "Warning: Encountered two children with the same key"


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/141872402/56a4c594-4d3e-4898-9b11-b87f7b7e2b11

**Review by XiaohanMeng — APPROVED** [review]
Hi, the buttons worked as intended!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113215317/4a569a7a-ad74-459d-af10-62080b9c37aa

File: src/components/BMDashboard/MaterialsList/RecordsModal.jsx [hunk 1]
```diff
@@ -74,22 +82,30 @@ export function Record({ record, recordType }) {
...
+                    </a>
+                  </td>
+                  <td>{moment(date).format('MM/DD/YY')}</td>
+                  <td>{status}</td>
+                </tr>
+              );
+            })
+          ) : (
+            <tr>
+              <td colSpan={4} style={{ fontWeight: 'bold' }}>
```
Comment (tdkent) [inline src/components/BMDashboard/MaterialsList/RecordsModal.jsx:104]: This should be changed to `colSpan={6}` so the text can correctly center align.

File: src/components/BMDashboard/MaterialsList/RecordsModal.jsx [hunk 2]
```diff
@@ -42,20 +42,28 @@ export function Record({ record, recordType }) {
...
-                    {`${data.createdBy.firstName} ${data.createdBy.lastName}`}
-                  </a>
-                </td>
-              </tr>
-            );
-          })}
+          {record?.updateRecord && record?.updateRecord.length ? (
+            record.updateRecord.map(data => {
+              return (
+                <tr key={data.date.toString() + data.createdBy._id}>
```
Comment (tdkent) [inline src/components/BMDashboard/MaterialsList/RecordsModal.jsx:48]: This key prop should be using the `_id` field to remove the dupe key prop error

File: src/components/BMDashboard/MaterialsList/RecordsModal.jsx [hunk 3]
```diff
@@ -74,22 +82,30 @@ export function Record({ record, recordType }) {
...
-                </td>
-                <td>{moment(date).format('MM/DD/YY')}</td>
-                <td>{status}</td>
-              </tr>
-            );
-          })}
+          {record?.purchaseRecord && record?.purchaseRecord.length ? (
+            record.map(({ date, status, brandPref, priority, quantity, requestedBy }) => {
+              return (
+                <tr key={date + requestedBy._id}>
```
Comment (tdkent) [inline src/components/BMDashboard/MaterialsList/RecordsModal.jsx:88]: This key prop should be using the _id field to remove the dupe key prop error

**Review by tdkent — CHANGES_REQUESTED** [review]
The functionality works as expected. However, there are a couple issues: 

1. The Purchase Record "no records" text is not center-aligned. On line 104 of **RecordsModal.jsx** the `colspan` of the `<td>` element should be changed from `4` to `6` (see my inline comment).

purch-rec-modal 

2. Your version of **RecordsModal.jsx** is also based off an older version of the file that needs to be merged with the current version of the file in `development`. This is why the duplicate key prop error is appearing in the console when the modals are used (see my inline comment), an issue you already fixed in #1813.

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**